pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/ComputeStep.java

```diff
@@ -195,23 +244,43 @@ private Object getAvroValue(Schema schema, Object value) {
     // Avro logical type conversion: https://avro.apache.org/docs/1.8.2/spec.html#Logical+Types
     switch (logicalType.getName()) {
       case "date":
-        validateLogicalType(value, schema.getLogicalType(), LocalDate.class);
-        LocalDate localDate = (LocalDate) value;
-        return (int) localDate.toEpochDay();
+        return getAvroData(value, schema.getLogicalType());
       case "time-millis":
-        validateLogicalType(value, schema.getLogicalType(), LocalTime.class);
-        LocalTime localTime = (LocalTime) value;
-        return (int) (localTime.toNanoOfDay() / 1000000);
+        return getAvroTimeMillis(value, schema.getLogicalType());
       case "timestamp-millis":
-        validateLogicalType(value, schema.getLogicalType(), OffsetDateTime.class);
-        OffsetDateTime offsetDateTime = (OffsetDateTime) value;
-        return offsetDateTime.toInstant().toEpochMilli();
+        return getAvroTimeTimestampMillis(value, schema.getLogicalType());
     }
 
     throw new IllegalArgumentException(
         String.format("Invalid logical type %s for value %s", schema.getLogicalType(), value));
   }
 
+  private Object getAvroTimeTimestampMillis(Object value, LogicalType logicalType) {
+    validateLogicalType(value, logicalType, Instant.class, Timestamp.class, LocalDateTime.class);
+    final Instant instant;
+    if (value instanceof Timestamp) {
+      instant = ((Timestamp) value).toInstant();
+    } else if (value instanceof LocalDateTime) {
+      instant = ((LocalDateTime) value).toInstant(ZoneOffset.UTC);
+    } else {
+      instant = (Instant) value;
+    }
+    return instant.toEpochMilli();
+  }
+
+  private Object getAvroData(Object value, LogicalType logicalType) {
+    validateLogicalType(value, logicalType, Date.class, LocalDate.class);
+    return (value instanceof LocalDate)
+        ? (int) ((LocalDate) value).toEpochDay()
+        : (int) (((Date) value).getTime() / TimeUnit.DAYS.toMillis(1));
```

Comment:
Need to extract `TimeUnit.DAYS.toMillis(1)` to a constant. Otherwise it's computed each time.